rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+  // The duration that `tbot` has been running for when it submitted this
+  // heartbeat.
+  google.protobuf.Duration uptime = 5;
+  // The currently configured join_method.
+  string join_method = 6;
+  // Indicates whether `tbot` is running in one-shot mode.
+  bool one_shot = 7;
+  // In future iterations, additional information can be submitted here.
+  // For example, the configuration of `tbot` or the health of individual
+  // outputs.
+}
+
+// BotInstanceStatusAuthentication contains information about a join or renewal.
+// Ths information is entirely sourced by the Auth Server and can be trusted.
+message BotInstanceStatusAuthentication {
+  // The timestamp that the join or renewal was authenticated by the Auth
+  // Server.
+  google.protobuf.Timestamp authenticated_at = 1;
+  // The join method used for this join or renewal.
+  string join_method = 2;
+  // The join token used for this join or renewal. This is only populated for
+  // delegated join methods as the value for `token` join methods is sensitive.
+  string join_token = 3;
+  // The metadata sourced from the join method.
+  google.protobuf.Struct metadata = 4;
+  // On each renewal, this generation is incremented. For delegated join
+  // methods, this counter is not checked during renewal. For the `token` join
+  // method, this counter is checked during renewal and the Bot is locked out if
+  // the counter in the certificate does not match the counter of the last
+  // authentication.
+  int32 generation = 5;
+  // The public key of the Bot instance.
+  // When authenticating a Bot instance, the full public key must be compared
+  // rather than just the fingerprint to mitigate pre-image attacks.
+  bytes public_key = 6;
+  // The fingerprint of the public key of the Bot instance.
+  string fingerprint = 7;
+}
+
+// BotInstanceStatus holds the status of a BotInstance.
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
+  // The name of the Bot that this instance is associated with.
+  string bot_name = 2;
+  // The initial authentication status for this bot instance.
+  BotInstanceStatusAuthentication initial_authentication = 3;
+  // The N most recent authentication status records for this bot instance.
+  repeated BotInstanceStatusAuthentication latest_authentications = 4;
+  // The initial heartbeat status for this bot instance.
+  BotInstanceStatusHeartbeat initial_heartbeat = 5;
+  // The N most recent heartbeats for this bot instance.
+  repeated BotInstanceStatusHeartbeat latest_heartbeats = 6;
+}
+```
+
+The name used for a BotInstance will be a concatenation of the Bot name and its
+unique identifier (UUID).
+
+When storing the BotInstance in the backend, the key will be:
+`bot_instances/{bot_name}/{uuid}`. This will allow for efficient listing
+of BotInstances for a given Bot.
+
+Like agent heartbeats, the BotInstance will expire after a period of inactivity.
+This avoids the accumulation of ephemeral BotInstances.
```

Comment:
Do we anticipate large numbers of bot instances at scale? Do we need to do something like variable rate heartbeats here, or is a fixed TTL reasonable?

Reply:
I'd guess a customer will eventually want to create a huge number of bot instances. What do you mean by variable rate, would jitter qualify?